File: cpp/main_tests.cpp
#include <iostream>
#include <cassert>
#include <string>
#include <vector>
#include <stdexcept>
#include <unordered_set>
#include <graph.hpp>
#include <graph_algorithms.hpp>
#include <graph_builders.hpp>

using namespace graph_lib;
using TestGraph = DirectedGraph<std::string, double>;
using UndirectedTestGraph = UndirectedGraph<std::string, double>;

// Helper function to safely run tests and catch "Not Implemented" errors
void runTest(const std::string& testName, std::function<void()> testFunc)
{
    std::cout << "Running " << testName << "... ";
    try {
        testFunc();
        std::cout << "\033[32m[PASSED]\033[0m\n"; // Green
    } catch (const std::logic_error& e) {
        std::cout << "\033[33m[PENDING]\033[0m (" << e.what() << ")\n"; // Yellow
    } catch (const std::exception& e) {
        std::cout << "\033[31m[FAILED]\033[0m (" << e.what() << ")\n"; // Red
    }
}

void testTraverseBFS()
{
    UndirectedTestGraph g;
    auto n1 = g.addNode("1"); auto n2 = g.addNode("2");
    auto n3 = g.addNode("3"); auto n4 = g.addNode("4");
    g.addArc(n1, n2, 0.0); g.addArc(n1, n3, 0.0); g.addArc(n2, n4, 0.0);

    std::vector<UndirectedTestGraph::NodePtr> visited_nodes;
    GraphTraversals<UndirectedTestGraph>::traverseBFS(g, n1, [&visited_nodes](UndirectedTestGraph::NodePtr n) {
        visited_nodes.push_back(n);
    });

    assert(visited_nodes.size() == 4 && "BFS should visit all 4 connected nodes exactly once.");
    assert(visited_nodes[0] == n1 && "BFS should start at the specified root node.");
}

void testTraverseDFS()
{
    UndirectedTestGraph g;
    auto n1 = g.addNode("1"); auto n2 = g.addNode("2");
    auto n3 = g.addNode("3"); auto n4 = g.addNode("4");
    g.addArc(n1, n2, 0.0); g.addArc(n2, n3, 0.0); g.addArc(n3, n4, 0.0);

    std::vector<UndirectedTestGraph::NodePtr> visited_nodes;
    GraphTraversals<UndirectedTestGraph>::traverseDFS(g, n1, [&visited_nodes](UndirectedTestGraph::NodePtr n) {
        visited_nodes.push_back(n);
    });

    assert(visited_nodes.size() == 4 && "DFS should visit all 4 connected nodes exactly once.");
    assert(visited_nodes[0] == n1 && "DFS should start at the specified root node.");
}

void testShortestPathBFS()
{
    UndirectedTestGraph g;
    auto n1 = g.addNode("1"); auto n2 = g.addNode("2");
    auto n3 = g.addNode("3"); auto n4 = g.addNode("4");
    g.addArc(n1, n2, 0.0); g.addArc(n1, n3, 0.0);
    g.addArc(n2, n4, 0.0); g.addArc(n3, n4, 0.0);

    auto path = GraphTraversals<UndirectedTestGraph>::findShortestPathBFS(g, n1, n4);
    assert(path.size() == 3 && "Shortest path from 1 to 4 should have 3 nodes");
    assert((path[1] == n2 || path[1] == n3) && "Middle node must be 2 or 3");
}

void testSpanningTreeDFS()
{
    UndirectedTestGraph g;
    auto n1 = g.addNode("1"); auto n2 = g.addNode("2");
    auto n3 = g.addNode("3"); auto n4 = g.addNode("4");
    g.addArc(n1, n2, 0.0); g.addArc(n1, n3, 0.0);
    g.addArc(n2, n4, 0.0); g.addArc(n3, n4, 0.0); // Cycle exists

    auto tree = GraphTraversals<UndirectedTestGraph>::buildSpanningTreeDFS(g, n1);
    assert(tree.getNodes().size() == 4 && "Spanning tree must include all nodes");
    assert(tree.getArcs().size() == 3 && "Spanning tree of 4 nodes must have exactly 3 arcs (no cycles)");
}

void testTopologicalSort()
{
    TestGraph g;
    auto n1 = g.addNode("CS101"); auto n2 = g.addNode("CS102");
    auto n3 = g.addNode("CS201"); auto n4 = g.addNode("CS301");

    g.addArc(n1, n2, 0.0); g.addArc(n2, n3, 0.0); g.addArc(n3, n4, 0.0);

    auto topo_order = GraphTopological<TestGraph>::getSort(g);
    assert(topo_order.size() == 4);
    assert(topo_order[0] == n1 && "First course must be CS101");
    assert(topo_order[3] == n4 && "Last course must be CS301");
}

void testTopologicalRanks()
{
    TestGraph g;
    auto n1 = g.addNode("CS101"); auto n2 = g.addNode("CS102");
    auto n3 = g.addNode("CS201"); auto n4 = g.addNode("CS301");

    g.addArc(n1, n2, 0.0); g.addArc(n2, n3, 0.0); g.addArc(n3, n4, 0.0);

    auto ranks = GraphTopological<TestGraph>::getRanks(g);
    assert(ranks[n1] == 0 && "CS101 should have rank 0");
    assert(ranks[n4] == 3 && "CS301 should have rank 3");
}

void testDijkstra()
{
    TestGraph g;
    auto nA = g.addNode("A"); auto nB = g.addNode("B"); auto nC = g.addNode("C");
    auto nD = g.addNode("D"); auto nE = g.addNode("E");

    g.addArc(nA, nB, 10.0); g.addArc(nA, nC, 3.0); g.addArc(nB, nD, 2.0);
    g.addArc(nC, nB, 1.0);  g.addArc(nC, nD, 8.0); g.addArc(nC, nE, 2.0);
    g.addArc(nE, nD, 9.0);

    auto spt = Dijkstra<TestGraph>::getMinimumPathsTree(g, nA);
    assert(spt.getNodes().size() == 5);
    assert(spt.getArcs().size() == 4);
}

void testKosarajuSCC()
{
    TestGraph g;
    auto nA = g.addNode("A"); auto nB = g.addNode("B");
    auto nC = g.addNode("C"); auto nD = g.addNode("D"); auto nE = g.addNode("E");

    g.addArc(nA, nB, 1.0); g.addArc(nB, nC, 1.0); g.addArc(nC, nA, 1.0);
    g.addArc(nC, nD, 1.0); // Bridge
    g.addArc(nD, nE, 1.0); g.addArc(nE, nD, 1.0);

    auto scc = GraphProperties<TestGraph>::computeStronglyConnectedComponents(g);
    assert(scc.size() == 2 && "Graph should have exactly 2 SCCs");
}

void testAStar()
{
    TestGraph g;
    auto nA = g.addNode("A"); auto nB = g.addNode("B"); auto nC = g.addNode("C");

    g.addArc(nA, nB, 1.0); g.addArc(nB, nC, 2.0); g.addArc(nA, nC, 10.0);

    auto heuristic = [](TestGraph::NodePtr n) -> double { return 0.0; }; 
    auto path = AStar<TestGraph>::findPath(g, nA, nC, heuristic);

    assert(path.size() == 3);
    assert(path[1] == nB);
}

void testBellmanFord()
{
    TestGraph g;
    auto n1 = g.addNode("1"); auto n2 = g.addNode("2"); auto n3 = g.addNode("3");

    g.addArc(n1, n2, 4.0); g.addArc(n2, n3, -2.0); g.addArc(n1, n3, 5.0);

    auto spt = BellmanFord<TestGraph>::getMinimumPathsTree(g, n1);
    assert(spt.getArcs().size() == 2);

    g.addArc(n3, n2, -5.0); // Create negative cycle
    bool caught = false;
    try { BellmanFord<TestGraph>::getMinimumPathsTree(g, n1); }
    catch (const std::exception&) { caught = true; }

    assert(caught && "Should detect negative cycle");
}

void testPrimKruskal()
{
    UndirectedTestGraph g;
    auto nA = g.addNode("A"); auto nB = g.addNode("B"); auto nC = g.addNode("C");

    g.addArc(nA, nB, 1.0); g.addArc(nB, nC, 2.0); g.addArc(nA, nC, 3.0);

    auto mst_prim = Prim<UndirectedTestGraph>::getMinimumSpanningTree(g, nA);
    auto mst_kruskal = Kruskal<UndirectedTestGraph>::getMinimumSpanningTree(g);

    assert(mst_prim.getArcs().size() == 2);
    assert(mst_kruskal.getArcs().size() == 2);
}

int main()
{
    std::cout << "==========================================\n";
    std::cout << "   GRAPH LIBRARY - STUDENT AUTOGRADER     \n";
    std::cout << "==========================================\n\n";

    runTest("Traverse BFS", testTraverseBFS);
    runTest("Traverse DFS", testTraverseDFS);
    runTest("Shortest Path (BFS)", testShortestPathBFS);
    runTest("Spanning Tree (DFS)", testSpanningTreeDFS);

    std::cout << "------------------------------------------\n";

    runTest("Topological Sort", testTopologicalSort);
    runTest("Topological Ranks", testTopologicalRanks);

    std::cout << "------------------------------------------\n";

    runTest("Dijkstra's Algorithm", testDijkstra);
    runTest("Kosaraju's SCC Algorithm", testKosarajuSCC);
    runTest("A* Algorithm", testAStar);
    runTest("Bellman-Ford Algorithm", testBellmanFord);
    runTest("Prim's Algorithm", testPrimKruskal);

    std::cout << "\n==========================================\n";
    return 0;
}

File: python/main_tests.py
import sys
from graph import DirectedGraph, UndirectedGraph
from graph_algorithms import Dijkstra, GraphTraversals, GraphProperties, GraphTopological, Prim, Kruskal, AStar, BellmanFord

def run_test(test_name: str, test_func: callable) -> None:
    """Helper function to safely run tests and catch NotImplemented errors."""
    print(f"Running {test_name.ljust(30)} ", end="")
    try:
        test_func()
        print("\033[32m[PASSED]\033[0m") # Green
    except NotImplementedError as e:
        print(f"\033[33m[PENDING]\033[0m ({str(e).split(' is ')[0]})") # Yellow
    except AssertionError as e:
        print(f"\033[31m[FAILED]\033[0m (Assertion Error: {e})") # Red
    except Exception as e:
        print(f"\033[31m[ERROR]\033[0m ({e})") # Red

def test_traverse_bfs() -> None:
    g = UndirectedGraph()
    n1 = g.add_node_by_value("1")
    n2 = g.add_node_by_value("2")
    n3 = g.add_node_by_value("3")
    n4 = g.add_node_by_value("4")
    g.add_arc(n1, n2, 0.0); g.add_arc(n1, n3, 0.0); g.add_arc(n2, n4, 0.0)

    visited_nodes = []
    def visit(n): visited_nodes.append(n)

    GraphTraversals.traverse_bfs(g, n1, visit)

    assert len(visited_nodes) == 4, "BFS should visit all 4 connected nodes exactly once."
    assert visited_nodes[0] == n1, "BFS should start at the specified root node."

def test_traverse_dfs() -> None:
    g = UndirectedGraph()
    n1 = g.add_node_by_value("1")
    n2 = g.add_node_by_value("2")
    n3 = g.add_node_by_value("3")
    n4 = g.add_node_by_value("4")
    g.add_arc(n1, n2, 0.0); g.add_arc(n2, n3, 0.0); g.add_arc(n3, n4, 0.0)

    visited_nodes = []
    def visit(n): visited_nodes.append(n)

    GraphTraversals.traverse_dfs(g, n1, visit)

    assert len(visited_nodes) == 4, "DFS should visit all 4 connected nodes exactly once."
    assert visited_nodes[0] == n1, "DFS should start at the specified root node."

def test_shortest_path_bfs() -> None:
    g = UndirectedGraph()
    n1 = g.add_node_by_value("1"); n2 = g.add_node_by_value("2")
    n3 = g.add_node_by_value("3"); n4 = g.add_node_by_value("4")
    g.add_arc(n1, n2, 0.0); g.add_arc(n1, n3, 0.0)
    g.add_arc(n2, n4, 0.0); g.add_arc(n3, n4, 0.0)

    path = GraphTraversals.find_shortest_path_bfs(g, n1, n4)
    assert len(path) == 3, "Shortest path from 1 to 4 should have 3 nodes"
    assert path[1] == n2 or path[1] == n3, "Middle node must be 2 or 3"

def test_spanning_tree_dfs() -> None:
    g = UndirectedGraph()
    n1 = g.add_node_by_value("1"); n2 = g.add_node_by_value("2")
    n3 = g.add_node_by_value("3"); n4 = g.add_node_by_value("4")
    g.add_arc(n1, n2, 0.0); g.add_arc(n1, n3, 0.0)
    g.add_arc(n2, n4, 0.0); g.add_arc(n3, n4, 0.0)

    tree = GraphTraversals.build_spanning_tree_dfs(g, n1)
    assert len(tree.nodes) == 4, "Spanning tree must include all nodes"
    assert len(tree.arcs) == 3, "Spanning tree of 4 nodes must have exactly 3 arcs"

def test_topological_sort() -> None:
    g = DirectedGraph()
    n1 = g.add_node_by_value("CS101"); n2 = g.add_node_by_value("CS102")
    n3 = g.add_node_by_value("CS201"); n4 = g.add_node_by_value("CS301")

    g.add_arc(n1, n2, 0.0); g.add_arc(n2, n3, 0.0); g.add_arc(n3, n4, 0.0)

    topo_order = GraphTopological.get_sort(g)
    assert len(topo_order) == 4
    assert topo_order[0] == n1, "First course must be CS101"
    assert topo_order[3] == n4, "Last course must be CS301"

def test_topological_ranks() -> None:
    g = DirectedGraph()
    n1 = g.add_node_by_value("CS101"); n2 = g.add_node_by_value("CS102")
    n3 = g.add_node_by_value("CS201"); n4 = g.add_node_by_value("CS301")

    g.add_arc(n1, n2, 0.0); g.add_arc(n2, n3, 0.0); g.add_arc(n3, n4, 0.0)

    ranks = GraphTopological.get_ranks(g)
    assert ranks[n1] == 0, "CS101 should have rank 0"
    assert ranks[n4] == 3, "CS301 should have rank 3"

def test_dijkstra() -> None:
    g = DirectedGraph()
    n_a = g.add_node_by_value("A"); n_b = g.add_node_by_value("B")
    n_c = g.add_node_by_value("C"); n_d = g.add_node_by_value("D")
    n_e = g.add_node_by_value("E")

    g.add_arc(n_a, n_b, 10.0); g.add_arc(n_a, n_c, 3.0); g.add_arc(n_b, n_d, 2.0)
    g.add_arc(n_c, n_b, 1.0);  g.add_arc(n_c, n_d, 8.0); g.add_arc(n_c, n_e, 2.0)
    g.add_arc(n_e, n_d, 9.0)

    spt = Dijkstra.get_minimum_paths_tree(g, n_a)
    assert len(spt.nodes) == 5
    assert len(spt.arcs) == 4

def test_kosaraju_scc() -> None:
    g = DirectedGraph()
    n_a = g.add_node_by_value("A"); n_b = g.add_node_by_value("B")
    n_c = g.add_node_by_value("C"); n_d = g.add_node_by_value("D")
    n_e = g.add_node_by_value("E")

    g.add_arc(n_a, n_b, 1.0); g.add_arc(n_b, n_c, 1.0); g.add_arc(n_c, n_a, 1.0)
    g.add_arc(n_c, n_d, 1.0) # Bridge
    g.add_arc(n_d, n_e, 1.0); g.add_arc(n_e, n_d, 1.0)

    scc = GraphProperties.compute_strongly_connected_components(g)
    assert len(scc) == 2, "Graph should have exactly 2 SCCs"

def test_a_star() -> None:
    g = DirectedGraph()
    na, nb, nc = g.add_node_by_value("A"), g.add_node_by_value("B"), g.add_node_by_value("C")
    g.add_arc(na, nb, 1.0); g.add_arc(nb, nc, 2.0); g.add_arc(na, nc, 10.0)

    path = AStar.find_path(g, na, nc, lambda n: 0.0)
    assert len(path) == 3
    assert path[1] == nb

def test_bellman_ford() -> None:
    g = DirectedGraph()
    n1, n2, n3 = g.add_node_by_value("1"), g.add_node_by_value("2"), g.add_node_by_value("3")
    g.add_arc(n1, n2, 4.0); g.add_arc(n2, n3, -2.0); g.add_arc(n1, n3, 5.0)

    spt = BellmanFord.get_minimum_paths_tree(g, n1)
    assert len(spt.arcs) == 2

    g.add_arc(n3, n2, -5.0)
    caught = False
    try:
        BellmanFord.get_minimum_paths_tree(g, n1)
    except RuntimeError:
        caught = True
    assert caught, "Should detect negative cycle"

def test_prim_kruskal() -> None:
    g = UndirectedGraph()
    na, nb, nc = g.add_node_by_value("A"), g.add_node_by_value("B"), g.add_node_by_value("C")
    g.add_arc(na, nb, 1.0); g.add_arc(nb, nc, 2.0); g.add_arc(na, nc, 3.0)

    mst_p = Prim.get_minimum_spanning_tree(g, na)
    mst_k = Kruskal.get_minimum_spanning_tree(g)

    assert len(mst_p.arcs) == 2 and len(mst_k.arcs) == 2

def main() -> None:
    print("==========================================")
    print("   GRAPH LIBRARY - STUDENT AUTOGRADER     ")
    print("==========================================\n")

    run_test("Traverse BFS", test_traverse_bfs)
    run_test("Traverse DFS", test_traverse_dfs)
    run_test("Shortest Path (BFS)", test_shortest_path_bfs)
    run_test("Spanning Tree (DFS)", test_spanning_tree_dfs)

    print("-" * 42)

    run_test("Topological Sort", test_topological_sort)
    run_test("Topological Ranks", test_topological_ranks)

    print("-" * 42)

    run_test("Dijkstra's Algorithm", test_dijkstra)
    run_test("Kosaraju's SCC Algorithm", test_kosaraju_scc)
    run_test("A* Algorithm", test_a_star)
    run_test("Bellman-Ford Algorithm", test_bellman_ford)
    run_test("Prim's & Kruskal's Algorithms", test_prim_kruskal)

    print("\n==========================================")

if __name__ == "__main__":
    main()